odoo/addons/base/transpiler/transpiler_js.py

```diff
@@ -0,0 +1,393 @@
+import re
+
+
+class TranspilerJS:
+    """
+    This class is what let us uses ES6 like JS in odoo.
+    Odoo expects all javascript to be divided into functions as "odoo.define(<module_name>, {...});".
+    If any file has the /** odoo-module */ on top of it, it will get processed by this class.
+    It performs several operations to get from ES6 syntax to the usual odoo one with minimal changes.
+    This is done on the fly, this not a pre-processing tool.
+
+    Caveat: This is done without a full parser, only using regex. One can only expect to cover as much edge cases
+    as possible with reasonable limitations. Also, this only changes imports and exports, so all JS features used in
+    the original source need to be supported by the browsers.
+    """
+
+
+    def __init__(self, content, url):
+        """
+        :param content: The original source code
+        :param url: The url of the file in the project
+        """
+        self.content = content
+        self.url = url
+        self.module_path = self.url_to_module_path(url)
+
+    def convert(self):
+        """
+        Transpile the code through a bunch of operations.
+        The order of the operations does sometimes matter.
+        :return: The transpiled source code
+        """
+        legacy_odoo_define = self.get_aliased_odoo_define_content()
+        self.convert_legacy_default_import()
+        self.convert_basic_import()
+        self.convert_default_import()
+        self.convert_star_import()
+        self.convert_unnamed_relative_import()
+        self.convert_from_export()
+        self.convert_star_from_export()
+        self.convert_relative_require()
+        self.remove_index()
+        self.convert_function_and_class_export()
+        self.convert_variable_export()
+        self.convert_list_export()
+        self.convert_default()
+        self.wrap_with_odoo_define()
+        if legacy_odoo_define:
+            self.content += legacy_odoo_define
+        return self.content
+
+    def url_to_module_path(self, url):
+        """
+        Odoo modules each have a name. (odoo.define("<the name>", function (require) {...});
+        It is used in to be required later. (const { something } = require("<the name>").
+        The transpiler transforms the url of the file in the project to this name.
+        It takes the module name and add a @ on the start of it, and map it to be the source of the static/src (or
+        static/tests) folder in that module.
+
+        in: web/static/src/one/two/three.js
+        out: @web/one/two/three.js
+        The module would therefore be defined and required by this path.
+
+        :param url: an url in the project
+        :return: a special path starting with @<module-name>.
+        """
+        result = re.match(r"""/?(?P<module>\w+)/[\w/]*static/(?P<type>src|tests)/(?P<url>[\w/]*)""", url)
```

Comment:
That seems to match `foo/barstatic/...` which I'm not sure is supposed to happen.